### PR TITLE
Feat/expose ipfs utils

### PR DESCRIPTION
**Pull request by JGAntunes**
Finishing the awesome work started by @interfect on #525.

I think I've addressed all your requests @diasdavid. Check it and let me know what you think.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -85,6 +85,9 @@ You can check the development status at:
   - [HTTP-API](#http-api)
   - [IPFS Core (use IPFS as a module in Node.js or in the Browser)](#ipfs-core-examples-use-ipfs-as-a-module)
     - [Create a IPFS node instance](#create-a-ipfs-node-instance)
+    - [Add a file](#add-a-file)
+    - [Retrieve a file](#retrieve-a-file)
```
Comment (diasdavid) [inline README.md]: I believe this is well demonstrated in the examples folder, we can trim the README by just pointing to those examples.

File: README.md [hunk 2]
```diff
@@ -252,6 +294,16 @@ Every IPFS instance also exposes the libp2p API at `ipfs.libp2p`. The formal int
...
 
+#### Domain data types
+
+A set of data types are exposed directly from the IPFS instance under `ipfs.types`. That way you're not required to import/require the following.
+
+* `ipfs.types.Buffer`
+* [`ipfs.types.PeerId`](https://github.com/libp2p/js-peer-id)
+* [`ipfs.types.PeerInfo`](https://github.com/libp2p/js-peer-info)
+* [`ipfs.types.multiaddr`](https://github.com/multiformats/js-multiaddr)
+* [`ipfs.types.multihash`](https://github.com/multiformats/js-multihash)
```
Comment (diasdavid) [inline README.md]: 👌🏽

File: examples/browser-script/README.md [hunk 3]
```diff
@@ -0,0 +1,13 @@
+# Use IPFS in the browser with `<script>` tags
```
Comment (diasdavid) [inline examples/browser-script/README.md]: call the folder 'browser-script-tag'

File: examples/libp2p-webrtc-star/README.md [hunk 4]
```diff
@@ -0,0 +1,14 @@
+# Robust Initialization and libp2p-webrtc-star Signaling
```
Comment (diasdavid) [inline examples/libp2p-webrtc-star/README.md]: Let's call this example 'how-to-use-webrtc-star'

File: README.md (hunk 2 above)
Comment (diasdavid) [inline README.md]: Also add CID https://github.com/ipld/js-cid

**Comment by JGAntunes** [comment]
@diasdavid I think I addressed everything.

File: README.md [hunk 5]
```diff
@@ -229,6 +232,10 @@ node.init({ emptyRepo: true, bits: 2048 }, (err) => {
 
 > We are working on making this init process better, see https://github.com/ipfs/js-ipfs/issues/556 for the discussion.
 
+More examples can be found in the [examples folder](./examples)
+
+> If you have built an example, please share it with the community by submitting a Pull Request to this repo!.
```
Comment (diasdavid) [inline README.md]: Let's remove this line, we don't necessarily want to have thousands of examples, instead we want to have a selected few that cover it all

File: examples/browser-script-tag/index.html [hunk 6]
```diff
@@ -0,0 +1,85 @@
+<!doctype html>
+<html>
+<head>
+  <title>IPFS in the Browser</title>
+  <script src="https://unpkg.com/ipfs/dist/index.min.js"></script>
+  <script type="text/javascript">
+    // Set this if you have a libp2p-webrtc-star server
+    // Something like
+    // const SIGNALING_SERVER = '/libp2p-webrtc-star/ip4/127.0.0.1/tcp/9090/ws/ipfs/'
+    const SIGNALING_SERVER = null
```
Comment (diasdavid) [inline examples/browser-script-tag/index.html]: Let's set this one with the sig server hosted by us https://github.com/libp2p/js-libp2p-webrtc-star#hosted-signalling-server

File: examples/browser-script-tag/index.html [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+    // Make an IPFS node
```
Comment (diasdavid) [inline examples/browser-script-tag/index.html]: s/Make/Create

File: examples/browser-script-tag/index.html [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    var ipfs = new Ipfs()
```
Comment (diasdavid) [inline examples/browser-script-tag/index.html]: s/ipfs/node

**Comment by JGAntunes** [comment]
Done @diasdavid :+1:

File: examples/browser-script-tag/index.html [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+
+    // Init a repo in the given IPFS node it if hasn't got one already. Calls the
+    // setup callback, passing the normal callback, after first initialization.
+    // Calls the normal callback directly after subsequent initializations. Calls
+    // the normal callback with an error parameter if there is an error.
```
Comment (diasdavid) [inline examples/browser-script-tag/index.html]: This sentence is super confusing, let's say instead:

```
Attempt to initialize the Repo, if one already exists, it uses that one.
```

File: examples/browser-script-tag/index.html [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+        // Go online and connect to things
+        ipfs.goOnline(() => {
+          console.log('Online status: ', ipfs.isOnline() ? 'online' : 'offline')
+          document.getElementById("status").innerHTML= 'Node status: ' + (ipfs.isOnline() ? 'online' : 'offline')
+          // TODO: Write your code here!
+          // Use methods like ipfs.files.add, ipfs.files.get, and so on in here
+          // Methods requiring buffers can use ipfs.Buffer
+        })
+      })
+    })
```
Comment (diasdavid) [inline examples/browser-script-tag/index.html]: Wanna try to improve this section 19-76, the code isn't super clear.

File: examples/browser-script-tag/index.html [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+  </script>
+</head>
+<body>
+  <h1>IPFS in the Browser</h1>
+  <p>This page creates an IPFS node in your browser and drops it into the global Javascript namespace as <em>ipfs</em>. Open the console to play around with it.</p>
```
Comment (diasdavid) [inline examples/browser-script-tag/index.html]: Add some suggestions of the commands that can be run. Like cat an hash that returns hello world :)

File: examples/how-to-use-webrtc-star/index.html [hunk 12]
```diff
@@ -0,0 +1,99 @@
+<!doctype html>
```
Comment (diasdavid) [inline examples/how-to-use-webrtc-star/index.html]: Sorry to be asking this now, but can you make this example a separate PR? I'm not super happy with the explanation provided and I know you are building on top of another contributor work, so don't feel the pressure to change it now. I can improve it :)

**Comment by diasdavid** [comment]
@JGAntunes almost there, one more round :)

File: src/core/index.js [hunk 13]
```diff
@@ -63,6 +73,9 @@ class IPFS {
     this.ping = components.ping(this)
     this.pubsub = components.pubsub(this)
 
+    // expose Buffer for browser applications
+    this.Buffer = Buffer
```
Comment (diasdavid) [inline src/core/index.js]: we can take it it, since we know have it under types.

File: examples/browser-script-tag/index.html (hunk 10 above)
Comment (JGAntunes) [inline examples/browser-script-tag/index.html]: Shall we make this compatible with ES5 or no need for it?

**Comment by JGAntunes** [comment]
Ok, I think this is it @diasdavid. I refactored a bit of the browser script tag example but tried to keep it simple without adding any unnecessary patterns given that is just an example, let me know what you think.

**Comment by diasdavid** [comment]
woot, thank you @JGAntunes :D

**Pull request merged**